File: RangeSumQuery-Immutable.py
class NumArray:

    def __init__(self, nums):
        """
        :type nums: List[int]
        """
        self.sums = [0] * (len(nums) + 1)
        for i in range(0, len(nums)):
            self.sums[i + 1] = self.sums[i] + nums[i];

    def sumRange(self, i, j):
        """
        :type i: int
        :type j: int
        :rtype: int
        """
        return self.sums[j + 1] - self.sums[i]


if __name__ == '__main__':
    nums = [int(num) for num in input().split()]
    arr = NumArray(nums)
    print(arr.sumRange(0, 2))
    print(arr.sumRange(2, 5))
    print(arr.sumRange(0, 5))

File: RangeSumQuery-Immutable.cpp
#include<iostream>
#include<vector>

using namespace std;

class NumArray {
private:
	vector<int> nums; 
public:
    NumArray(vector<int> nums) {
        this->nums = nums;
    }
    
    int sumRange(int i, int j) {
        int sum = 0;
        for(int k = i;k <= j;k++) sum += nums[k];
        return sum;
    }
};

int main() {
	int n;
	cin>>n;
	
	vector<int> nums;
	for(int i = 0;i < n;i++) {
		int num;
		cin>>num;
		nums.push_back(num);
	}
	
	NumArray *arr = new NumArray(nums);
	
	cout<<arr->sumRange(0, 2)<<endl;
	cout<<arr->sumRange(2, 5)<<endl;
	cout<<arr->sumRange(0, 5)<<endl;
	
	return 0; 
}
